File: Amazon Interviews/21. Merge Two Sorted Lists.py
class Solution:
    def mergeTwoList(self, l1, l2):
        dummy = ListNode(0)
        head = dummy

        while l1 and l2:
            if l1.val < l2.val:
                head.next = l1
                head = head.next
                l1 = l1.next
            else:
                head.next = l2
                head = head.next
                l2 = l2.next

        head.next = l1 if l1 else l2

        return dummy.next

# o(n) running time
# o(1) space complexity

File: Leetcode First 200 JS solution/21. Merge Two Sorted Lists.js
var mergeTwoLists = function(l1, l2) {

    let dummy = new ListNode(1, l1);
    let p = dummy;

    while (l1 && l2) {
        if (l1.val <= l2.val) {
            p.next = l1;
            p = p.next;
            l1 = l1.next
        } else {
            p.next = l2;
            p = p.next;
            l2 = l2.next;
        }
    }

    if (l1) p.next = l1;
    if (l2) p.next = l2;
    return dummy.next;
};

// o(n+m)


